### PR TITLE
Move parts of emscripten exception handling to native code. NFC

### PR DESCRIPTION
**Pull request by sbc100**
Specifically, this change moves the allocation and reference counting
functions.

This saves on code size but more importantly reduces the number and of
complexity of imports/exports, which in turn helps with the wasm64 work
I've been doing.

### CODE REVIEW

File: system/lib/libcxxabi/src/cxa_emscripten.cpp
Comment (sbc100) [reply system/lib/libcxxabi/src/cxa_emscripten.cpp:23]: Sorry, yes, most of the code in the file is just copied from `src/cxa_exception.cpp`... I should have written that.

File: system/lib/libcxxabi/src/cxa_emscripten.cpp
Comment (sbc100) [reply system/lib/libcxxabi/src/cxa_emscripten.cpp:8]: The DEBUG macro in this file is useful for printing debugging information but here I define it away to nothing.   For debugging purposes one would change this 1 to a 0.

File: tests/test_core.py
Comment (sbc100) [reply tests/test_core.py:1213]: No, this test actually relies on the output the `EXCEPTION_DEBUG`.. for better or worse, I'm just moving this setting down a few lines so its only used then its actually needed.

File: system/lib/libcxxabi/src/cxa_emscripten.cpp
Comment (kripken) [reply system/lib/libcxxabi/src/cxa_emscripten.cpp]: I see, thanks. Yes, a comment mentioning that sgtm.

File: system/lib/libcxxabi/src/cxa_exception_emscripten.cpp
Comment (kripken) [inline system/lib/libcxxabi/src/cxa_exception_emscripten.cpp:15]: Perhaps instead of `1` this could be a name like `EMSCRIPTEN_DEBUG`? The `1` looks like a mistake, to my eye at least, but I do understand what you said about it. lgtm either way though.

**Comment by kripken** [comment]
Diffing the cpp file to the file it is adapted from, I see stuff like this:
```diff
 //  Allocate a __cxa_exception object, and zero-fill it.
 //  Reserve "thrown_size" bytes on the end for the user's exception
 //  object. Zero-fill the object. If memory can't be allocated, call
 //  std::terminate. Return a pointer to the memory to be used for the
 //  user's exception object.
 void *__cxa_allocate_exception(size_t thrown_size) _NOEXCEPT {
     size_t actual_size = cxa_exception_size_from_exception_thrown_size(thrown_size);
 
-    // Allocate extra space before the __cxa_exception header to ensure the
-    // start of the thrown object is sufficiently aligned.
-    size_t header_offset = get_cxa_exception_offset();
     char *raw_buffer =
-        (char *)__aligned_malloc_with_fallback(header_offset + actual_size);
+        (char *)__aligned_malloc_with_fallback(actual_size);
     if (NULL == raw_buffer)
         std::terminate();
     __cxa_exception *exception_header =
-        static_cast<__cxa_exception *>((void *)(raw_buffer + header_offset));
+        static_cast<__cxa_exception *>((void *)(raw_buffer));
     ::memset(exception_header, 0, actual_size);
     return thrown_object_from_cxa_exception(exception_header);
 }
```
Why do we not have the `header_offset`? Documenting such changes would be good.

The other part of the diff that looks unclear to me is this:
```diff
 void __cxa_decrement_exception_refcount(void *thrown_object) _NOEXCEPT {
     if (thrown_object != NULL )
     {
         __cxa_exception* exception_header = cxa_exception_from_thrown_object(thrown_object);
-        if (std::__libcpp_atomic_add(&exception_header->referenceCount, size_t(-1)) == 0)
+        DEBUG("DEC: %p refcnt=%zu rethrown=%d\n", thrown_object, exception_header->referenceCount, exception_header->rethrown);
+        assert(exception_header->referenceCount > 0);
+        if (std::__libcpp_atomic_add(&exception_header->referenceCount, size_t(-1)) == 0 && !exception_header->rethrown)
         {
+            DEBUG("DEL: %p\n", thrown_object);
             if (NULL != exception_header->exceptionDestructor)
                 exception_header->exceptionDestructor(thrown_object);
             __cxa_free_exception(thrown_object);
         }
     }
 }
```
We add `&& !exception_header->rethrown` but I'm not sure why?

**Comment by sbc100** [comment]
Yes, those deltas are deliberate, and made to match the JS implementation.  Honestly I'm not totally sure why they exist, but they match the existing JS code.

**Comment by sbc100** [comment]
Added more comments

**Review by kripken — APPROVED** [review]
Ok, sgtm to match the old JS behavior even if we don't fully understand it.

**Comment by RReverser** [comment]
Oh I started working on something similar but then decided to search PRs just in case, and glad I did. There seems to be a lot of conflicts - mainly in tests - but otherwise seems like this was in mergeable state?

**Comment by sbc100** [comment]
I landed a smaller version of this patch in #18292.   I'm not sure its worth revisiting this larger change.   Would be good to have but not sure if its high enough priority given that it works, and we have native exception handling coming down the pipe soon.

**Comment by RReverser** [comment]
> I'm not sure its worth revisiting this larger change. Would be good to have but not sure if its high enough priority given that it works, and we have native exception handling coming down the pipe soon.

Unfortunately, one doesn't contradict the other. I still need to work on something like this for project I'm working on at the moment, because neither Wasmtime nor Wasmer support Wasm exceptions yet, so the only chance for STANDALONE_WASM is manually shimming out Emscripten exceptions in the custom runtime we have.

Moving parts of the Emscripten EH that just modify C++ exception structures from JS to C++ would make this much easier, as in the end I'd only need to shim out `invoke_*` and `__cxa_throw` (and maybe few other small things) rather than the whole API.

**Comment by RReverser** [comment]
I guess, looking at the number of conflicts, I should just try & start anew after all...

**Comment by RReverser** [comment]
I'll try to rebase it first after all.

**Comment by sbc100** [comment]
Sounds good.  I would still love to see this land if you can make it happen.

File: system/lib/libcxxabi/src/cxa_exception_emscripten.cpp
Comment (sbc100) [inline system/lib/libcxxabi/src/cxa_exception_emscripten.cpp:1]: This filename looks wrong.

File: system/lib/libcxxabi/src/cxa_exception_emscripten.cpp
Comment (sbc100) [inline system/lib/libcxxabi/src/cxa_exception_emscripten.cpp:49]: These utilities seem to be duplicated here and in the new utils file.

File: test/other/metadce/test_metadce_cxx_mangle.exports
Comment (sbc100) [inline test/other/metadce/test_metadce_cxx_mangle.exports:8]: Conflict markers here.

**Comment by RReverser** [comment]
@sbc100 FWIW this is WIP, I just wanted to push as I got far enough in my conflict resolution & new changes that I didn't want to lose.

File: system/lib/libcxxabi/src/cxa_exception_emscripten.cpp
Comment (RReverser) [reply system/lib/libcxxabi/src/cxa_exception_emscripten.cpp:49]: Yeah, just like in cxa_exception they were originally borrowed from, they're static local helpers everywhere.

**Comment by RReverser** [comment]
I'll continue working on this tomorrow. For now I managed to rebase most of this stuff, although had to deal with new helpers exposed to JS that were added since this PR.

Since they're shared between backends, I've moved them to a new `.cpp` file together with (yet another) copy of used `static` function helpers.

Ideal end state I think would be to reuse `cxa_exception.cpp` between both backends with some `ifdef` sprinkled to exclude bits implemented in JS, but that would be out of scope of this particular PR - for now just trying to get it to do what it did originally.

File: emcc.py
Comment (RReverser) [inline emcc.py:2798]: Also had to add `__cxa_decrement_exception_refcount` here to make some tests pass; I'm not sure why it wasn't necessary for the original PR but became now - maybe I'm missing a better place where it should be added, or maybe something changed in tests.

File: system/lib/libcxxabi/src/cxa_exception_emscripten.cpp
Comment (RReverser) [reply system/lib/libcxxabi/src/cxa_exception_emscripten.cpp:1]: That's what you called it in your version: https://github.com/emscripten-core/emscripten/commit/3efb14451275d1806a0cb51ba6c9efa011a79560#diff-b09ab8d2f1ceaab1178f47623c925047cfec30a707e60e95d2ded9f3b7e70a14 (this is the link to your original commit)

Do you want me to change it?

Comment (sbc100) [reply system/lib/libcxxabi/src/cxa_exception_emscripten.cpp:1]: I mean the comment is wrong..  but yet it was wrong in my original PR :)

Comment (RReverser) [reply system/lib/libcxxabi/src/cxa_exception_emscripten.cpp:1]: Do you mean just the filename or is there anything else wrong here? I didn't pay much attention to parts that rebased cleanly, but happy to fix them too while I'm here, just let me know what 😅

Comment (sbc100) [reply system/lib/libcxxabi/src/cxa_exception_emscripten.cpp:1]: I just mean the name of the file in comment doesn't match the name of the file itself.

**Comment by sbc100** [comment]
`EXPORT_EXCEPTION_HANDLING_HELPERS` is something that is used to export helper functions for inspecting thrown excetpions from JS.  This is above and beyond what is need to use exceptions with your app (i.e. throw and catch them).   They are two different things, one is required (to make use of exceptions) and one is optional (to be able to inspect them from the JS side).

**Comment by RReverser** [comment]
Sounds good. Other than that, this should be ready for review (again).

That Firefox failure seems unrelated.

**Comment by RReverser** [comment]
I also want to move more things to C++, but I'll do that in a separate PR.

File: test/test_core.py
Comment (sbc100) [inline test/test_core.py:4]: Not sure why this was deleted... revert?

File: test/test_core.py
Comment (sbc100) [inline test/test_core.py:1271]: Was there a reason for this change?  (Maybe I did this, but I can't remember why).

**Review by sbc100 — COMMENTED** [review]
lgtm if tests pass.

Looks like I can't actually approve it because its my PR.. so maybe you will have to.

File: test/test_core.py
Comment (RReverser) [reply test/test_core.py:1271]: Yeah you did, but no, I don't know why.

File: test/test_core.py
Comment (RReverser) [reply test/test_core.py:4]: Hm yeah looks like unfortunate rebase artifact.

File: src/jsifier.js
Comment (RReverser) [reply src/jsifier.js:298]: It was yeah.

Comment (sbc100) [reply src/jsifier.js:298]: Can we revert this part unless we know why its needed?

File: test/test_core.py
Comment (RReverser) [reply test/test_core.py:1271]: Looks like you discussed this above:
image

Comment (sbc100) [reply test/test_core.py:1271]: That person sounds very logical.  I can't argue with them.

File: src/jsifier.js
Comment (RReverser) [reply src/jsifier.js]: Done.

**Comment by sbc100** [comment]
lgtm

File: system/lib/libcxxabi/src/cxa_exception_js_utils.cpp
Comment (sbc100) [inline system/lib/libcxxabi/src/cxa_exception_js_utils.cpp:88]: Before you land, can you verify that these functions did not change since they were from one file to another?   i.e. are these function 100% unchanged?

Comment (RReverser) [reply system/lib/libcxxabi/src/cxa_exception_js_utils.cpp:88]: Hm yeah I think I saw some recent exception-related PRs, worth double-checking...

Comment (sbc100) [reply system/lib/libcxxabi/src/cxa_exception_js_utils.cpp:88]: Can we land this now?

Comment (RReverser) [reply system/lib/libcxxabi/src/cxa_exception_js_utils.cpp:88]: I'll rebase now.

**Comment by sbc100** [comment]
This seems to have broken some LTO tests: https://logs.chromium.org/logs/emscripten-releases/buildbucket/cr-buildbucket/8784015253103176065/+/u/Emscripten_testsuite__LTO_/stdout

e.g:

```
RuntimeError: function signature mismatch
    at __cxa_decrement_exception_refcount (<anonymous>:wasm-function[1464]:0x2173e)
    at /usr/local/google/home/sbc/dev/wasm/emscripten/out/test/test_exceptions_allowed_uncaught.js:719:22
    at ___cxa_end_catch (/usr/local/google/home/sbc/dev/wasm/emscripten/out/test/test_exceptions_allowed_uncaught.js:1238:7)
    at invoke_v (/usr/local/google/home/sbc/dev/wasm/emscripten/out/test/test_exceptions_allowed_uncaught.js:4570:29)
    at <anonymous>:wasm-function[86]:0x6b2f
    at <anonymous>:wasm-function[141]:0x780a
````

**Comment by RReverser** [comment]
Ugh I was hoping Emscripten's own CI checks would be enough, and they passed. Can you revert please?

**Comment by sbc100** [comment]
I'm investigating now.  Will revert if I can't find a quick fix.

**Comment by sbc100** [comment]
(sadly we don't have time to run all tests in all configurations during CI, so we run some only the emscripten-releases waterfall: https://ci.chromium.org/p/emscripten-releases/g/main/console)

**Comment by RReverser** [comment]
From a quick look, the signature seems the same as before despite moving func from JS to C++. I'm curious why it complains...

**Comment by sbc100** [comment]
Its particularly interesting that it only happens for LTO..

**Comment by sbc100** [comment]
I think I have a fix.

**Comment by RReverser** [comment]
Could it be related to _LIBCXXABI_NO_CFI?

**Comment by dschuff** [comment]
This is still causing a failure in the LLVM testsuite: https://logs.chromium.org/logs/emscripten-releases/buildbucket/cr-buildbucket/8783849429524138417/+/u/LLVM_testsuite/stdout

```
wasm-ld: error: lib/libgtest.a(gtest-all.cc.o): undefined symbol: __cxa_free_exception
em++: error: '/b/s/w/ir/x/w/install/bin/wasm-ld -o MicroBenchmarks/libs/benchmark/test/perf_counters_gtest.wasm -L/b/s/w/ir/cache/builder/emscripten-releases/build/llvmtest-out MicroBenchmarks/libs/benchmark/test/CMakeFiles/perf_counters_gtest.dir/perf_counters_gtest.cc.o MicroBenchmarks/libs/benchmark/src/libbenchmark.a lib/libgmock_main.a lib/libgmock.a lib/libgtest.a -L/b/s/w/ir/x/w/install/emscripten/cache/sysroot/lib/wasm32-emscripten -lGL -lal -lhtml5 -lstubs -lc -ldlmalloc -lcompiler_rt -lc++-noexcept -lc++abi-noexcept -lsockets -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr --allow-undefined-file=/b/s/w/ir/x/t/tmp2wer4hv6.undefined --strip-debug --export-if-defined=main --export-if-defined=__start_em_asm --export-if-defined=__stop_em_asm --export-if-defined=__start_em_lib_deps --export-if-defined=__stop_em_lib_deps --export-if-defined=__start_em_js --export-if-defined=__stop_em_js --export-if-defined=__main_argc_argv --export-if-defined=fflush --export=stackSave --export=stackRestore --export=stackAlloc --export=__errno_location --export=__get_temp_ret --export=__set_temp_ret --export=__funcs_on_exit --export=__wasm_call_ctors --export=__cxa_is_pointer_type --export=malloc --export=__cxa_increment_exception_refcount --export=setThrew --export=__cxa_decrement_exception_refcount --export=__cxa_can_catch --export=setTempRet0 --export-table -z stack-size=131072 --initial-memory=1073741824 --no-entry --max-memory=1073741824 --global-base=1024' failed (returned 1)
```

**Comment by dschuff** [comment]
I have it reproing locally too, let me know if you want more detail or the files.

**Comment by sbc100** [comment]
I'm guessing that test is building with the default settings?  (i.e. `DISABLE_EXCEPTION_CATCHING=1` + `DISABLE_EXCEPTION_THROWING=0`).. hense the linking with `-lc++-noexcept -lc++abi-noexcept`.     Are you sure it was this change that caused the failure to start?

**Comment by sbc100** [comment]
I'm guessing the change to system/lib/libcxxabi/src/cxa_noexception.cpp that was part of this change needs to have __cxa_free_exception as well as  the __cxa_allocate_exception

**Comment by sbc100** [comment]
Yup that looks like it the problem.. the sad thing is I have no idea why non of our core tests required that.  Any idea what `gtest-all.cc.o` is doing to depend on that symbol?  (so we can add a test?)

**Comment by dschuff** [comment]
According to https://itanium-cxx-abi.github.io/cxx-abi/abi-eh.html#cxx-destroy it seems like `__cxa_end_catch` should usually call `__cxa_free_exception` (and indeed this seems to be the case in cxa_exception.cpp where it's called via `cxa_decrement_exception_refcount` and this is mirrored in cxa_exception_emscripten.cpp. So it seems that basically anything with a catch block should depend on it?

**Comment by dschuff** [comment]
Oh wait, I see what you're saying. If catching is disabled, then why are we still getting this dependency? Yeah I don't quite get it yet either.

**Comment by sbc100** [comment]
In emscripten we default to throwing being enabled but catching being disabled.. which is a little unusual.

**Comment by sbc100** [comment]
I don't see `__cxa_free_exception` in the source code, so I assume you don't mean they the are actually calling it directly from their source, but that clang is generating a call to it?

**Comment by sbc100** [comment]
I looks like it must be generated here: https://github.com/llvm/llvm-project/blob/64ae7669a7cefa12ea1117680cf9bfb5c3f6084c/clang/lib/CodeGen/CGException.cpp#L380-L388

Maybe we can add a test that simply calls it directly?  Or figure out how to trigger the above code indirectly in C++?

**Comment by dschuff** [comment]
Yeah, if a generated call is possible, then a test that calls it directly probably makes sense.

**Pull request merged**
